interpreter/cling/lib/Interpreter/ValuePrinter.cpp

```diff
@@ -344,8 +344,452 @@ static void StreamValue(llvm::raw_ostream& o, const void* V,
+      case clang::BuiltinType::Long: // intentional fall through
+      case clang::BuiltinType::LongLong: {
+        std::ostringstream strm;
+        strm << V.getLL();
+        return strm.str();
+      }
+        break;
+      case clang::BuiltinType::UChar: // intentional fall through
+      case clang::BuiltinType::UShort: // intentional fall through
+      case clang::BuiltinType::UInt: // intentional fall through
+      case clang::BuiltinType::ULong: // intentional fall through
+      case clang::BuiltinType::ULongLong: {
+        std::ostringstream strm;
+        strm << V.getULL();
+        return strm.str();
+      }
+        break;
+      case clang::BuiltinType::Float: {
+        std::ostringstream strm;
+        strm << V.getFloat();
+        return strm.str();
+      }
+        break;
+      case clang::BuiltinType::Double: {
+        std::ostringstream strm;
+        strm << V.getDouble();
+        return strm.str();
+      }
+        break;
+      case clang::BuiltinType::LongDouble: {
+        std::ostringstream strm;
+        strm << V.getLongDouble();
+        return strm.str();
+      }
+        break;
+      default:
+        std::ostringstream strm;
+        strm << V.getPtr();
+        return strm.str();
+        break;
+    }
+  }
+  else if (Ty->isIntegralOrEnumerationType()) {
+    std::ostringstream strm;
+    strm << V.getLL();
+    return strm.str();
+  }
+  else if (Ty->isFunctionType()) {
+    std::ostringstream strm;
+    strm << (const void *) &V;
+    return strm.str();
+  } else if (Ty->isPointerType() || Ty->isReferenceType()
+             || Ty->isArrayType()) {
+    std::ostringstream strm;
+    strm << V.getPtr();
+    return strm.str();
+  }
+  else {
+    // struct case.
+    std::ostringstream strm;
+    strm << V.getPtr();
+    return strm.str();
+  }
+}
+
+static std::string getCastValueString(const Value &V) {
+  std::ostringstream strm;
+  clang::ASTContext &C = V.getASTContext();
+  clang::QualType Ty = V.getType().getDesugaredType(C).getNonReferenceType();
+  std::string type = cling::utils::TypeName::GetFullyQualifiedName(Ty, C);
+  std::ostringstream typeWithOptDeref;
+  std::ostringstream suffix;
+
+  if (llvm::dyn_cast<clang::BuiltinType>(Ty.getCanonicalType())){
+    typeWithOptDeref << "(" << type << ")";
+  } else if (Ty->isPointerType()) {
+    if (Ty->getPointeeType()->isCharType()) {
+      typeWithOptDeref << "(" << type << ")";
+    } else {
+      typeWithOptDeref << "(void*)";
+    }
+  } else if (Ty->isArrayType()) {
+    const clang::ArrayType* ArrTy = Ty->getAsArrayTypeUnsafe();
+    clang::QualType ElementTy = ArrTy->getElementType();
+    if (Ty->isConstantArrayType()) {
+      const clang::ConstantArrayType* CArrTy = C.getAsConstantArrayType(Ty);
+      const llvm::APInt& APSize = CArrTy->getSize();
+      size_t size = (size_t)APSize.getZExtValue();
+
+      typeWithOptDeref << "(" << ElementTy.getAsString() << "(&)[" << size << "])*(void*)";
```

Comment:
Can you add a comment what the result of this looks like? I.e. how to reference the array in C++. For the uneducated reader ;-)